File: computer3/Fall Term/Solutions/Class05_Oct_09/lvl3_is_it_a_tree.java
import java.util.*;
import java.io.*;
public class homework {
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static PrintWriter pr = new PrintWriter(new BufferedWriter(new OutputStreamWriter(System.out)));
	static StringTokenizer st;
	public static void main(String[] args) throws IOException{
		int [][] adj = new int[4][4];
		int sum = 0; boolean disconnect = false;
		for(int i=0; i<4; i++) {
			int row = 0;
			for(int j=0; j<4; j++) {
				adj[i][j] = readInt();
				sum += adj[i][j];
				row += adj[i][j];
			}
			if(row == 0) disconnect = true;
		}
		if(sum == 6 && !disconnect) System.out.println("Yes");
		else System.out.println("No");
	}
	static String next () throws IOException {
		while (st == null || !st.hasMoreTokens())
			st = new StringTokenizer(br.readLine().trim());
		return st.nextToken();
	}
	static long readLong () throws IOException {
		return Long.parseLong(next());
	}
	static int readInt () throws IOException {
		return Integer.parseInt(next());
	}
	static double readDouble () throws IOException {
		return Double.parseDouble(next());
	}
	static char readCharacter () throws IOException {
		return next().charAt(0);
	}
	static String readLine () throws IOException {
		return br.readLine().trim();
	}
}

File: computer3/Fall Term/Solutions/Class05_Oct_09/lvl3_is_it_a_tree.cpp
#include <stdio.h>
int adj[4][4], vis[4], i, j, cnt;
int main(){
    for(i=0; i<4; i++)
        for(j=0; j<4; j++){
            scanf("%d", &adj[i][j]);
            cnt += adj[i][j];
        }
    if(cnt!=6) {printf("No\n"); return 0;}
    for(i=0; i<4; i++){
        cnt = 0;
        for(j=0; j<4; j++){
            cnt += adj[i][j];
        }
        if(!cnt) {printf("No\n"); return 0;}
    }
    printf("Yes\n"); return 0;
}